src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+                assert_eq!(expected_point.to_affine(), actual_point);
+            }
+        }
+
+        Ok((EccPoint::from_coordinates_unchecked(x_a.0, y_a), zs_sum))
+    }
+
+    #[allow(clippy::type_complexity)]
+    // Hash a message piece containing `piece.length` number of `K`-bit words.
+    fn hash_piece(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        offset: usize,
+        piece: &<Self as SinsemillaInstructions<
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::MessagePiece,
+        x_a: X<pallas::Base>,
+        y_a: Y<pallas::Base>,
+    ) -> Result<
+        (
+            X<pallas::Base>,
+            Y<pallas::Base>,
+            Vec<CellValue<pallas::Base>>,
+        ),
+        Error,
+    > {
+        let config = self.config().clone();
+
+        // Selector assignments
+        {
+            // Enable `q_sinsemilla1` selector on every row.
+            for row in 0..piece.num_words() {
+                config.q_sinsemilla1.enable(region, offset + row)?;
+            }
+
+            // Set `q_sinsemilla2` fixed column to 1 on every row but the last.
+            for row in 0..(piece.num_words() - 1) {
+                region.assign_fixed(
+                    || "q_s2 = 1",
+                    config.q_sinsemilla2,
+                    offset + row,
+                    || Ok(pallas::Base::one()),
+                )?;
+            }
+
+            // Set `q_sinsemilla2` fixed column to 0 on the last row.
+            region.assign_fixed(
+                || "q_s2 = 1",
+                config.q_sinsemilla2,
+                offset + piece.num_words() - 1,
+                || Ok(pallas::Base::zero()),
+            )?;
+        }
+
+        // Message piece as K * piece.length bitstring
+        let bitstring: Option<Vec<bool>> = piece.field_elem().map(|value| {
+            value
+                .to_le_bits()
+                .into_iter()
+                .take(sinsemilla::K * piece.num_words())
+                .collect()
+        });
+
+        let words: Option<Vec<u32>> = bitstring.map(|bitstring| {
+            bitstring
+                .chunks_exact(sinsemilla::K)
+                .map(|word| lebs2ip_k(word))
+                .collect()
+        });
+
+        // Get (x_p, y_p) for each word. We precompute this here so that we can use `batch_normalize()`.
+        let generators_projective: Option<Vec<pallas::Point>> = words
+            .clone()
+            .map(|words| words.iter().map(|word| get_s_by_idx(*word)).collect());
+        let generators: Option<Vec<(pallas::Base, pallas::Base)>> =
+            generators_projective.map(|generators_projective| {
+                let mut generators = vec![pallas::Affine::default(); generators_projective.len()];
+                pallas::Point::batch_normalize(&generators_projective, &mut generators);
```

Comment:
Rather than batch-normalizing the generators each time we use them, we should batch normalize them once and for all when the table is computed.

Reply:
Is it worth just precomputing and storing the generators in `primitives::sinsemilla::constants`?

Reply:
Probably, but let's try that after this PR.

Reply:
Opened an issue for this https://github.com/zcash/orchard/issues/124